File: LeetCode/cpp/313.cpp
class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int> &primes) {
        int m = primes.size();
        vector<int> factor(primes.size(), 0), res(n, 1);
        for (int i = 1; i < n; ++i) {
            res[i] = INT_MAX;
            for (int j = 0; j < m; ++j)
                res[i] = min(res[i], primes[j] * res[factor[j]]);
            for (int j = 0; j < m; ++j)
                if (primes[j] * res[factor[j]] == res[i])
                    ++factor[j];
        }
        return res.back();
    }
};

File: LeetCode/python3/313.py
class Solution:
    def nthSuperUglyNumber(self, n: int, primes: List[int]) -> int:
        m = len(primes)
        factor, res = [0 for _ in range(m)], [1 for _ in range(n)]
        for i in range(1, n):
            res[i] = sys.maxsize
            for j in range(m):
                res[i] = min(res[i], primes[j] * res[factor[j]])
            for j in range(m):
                if primes[j] * res[factor[j]] == res[i]:
                    factor[j] += 1
        return res[-1]
